Implement bucketed tables for Presto on Spark

arhimondr opened the pull request:
```
== NO RELEASE NOTE ==
```

arhimondr commented:
Milestone for #13856

wenleix commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java:
```diff
@@ -70,4 +70,12 @@ public BucketFunction getBucketFunction(ConnectorTransactionHandle transactionHa
+    @Override
+    public int getBucketCount(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
+    {
+        Set<Node> nodes = nodeManager.getRequiredWorkerNodes();
```
One theoretic caveat here is the number of buckets, returned by `getBucketNodeMap` and `getBucketCount` can be different if the number of nodes changed during the two method calls. Since `nodeManager.getRequiredWorkerNodes()` will return different list.

However, this would only happen for 3 testing connector: BlackHole, TPCDS, TPCH as the "partitioning handle" is somewhat artificial in these connectors. For "real" connector partitioning handle, the number of buckets should be a constant property in the partitioning handle, thus. it's not a problem.

I think the right long term solution is to allow connector partitioning handle to return the number of buckets :). So the current implementation looks OK.

BTW: can we just throw in `getBucketCount` for BlackHole, TPCDS and TPCH connector?

arhimondr replied:
> One theoretic caveat here is the number of buckets, returned by getBucketNodeMap and getBucketCount can be different if the number of nodes changed during the two method calls. Since nodeManager.getRequiredWorkerNodes() will return different list.

Yeah, that is in theory possible. However it is also possible that `getBucketNodeMap` called multiple times may return different bucket count. But I agree, that is a little brittle. 

> BTW: can we just throw in getBucketCount for BlackHole, TPCDS and TPCH connector?

I would prefer to keep it as is, as it is sometimes useful to have TPC-DC and TPC-H connectors enabled in Presto on Spark.

wenleix replied:
> Yeah, that is in theory possible. However it is also possible that getBucketNodeMap called multiple times may return different bucket count. But I agree, that is a little brittle.

Right. Long long ago (before grouped execution), `getBucketNodeMap` is only called once in execution. Now we already called once in `PlanFragmenter`, once in execution. 

Anyhow. :)

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/SerializedTaskSource.java:
```diff
@@ -0,0 +1,34 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class SerializedTaskSource
```
nit: maybe call it `SerializedPrestoSparkTaskSource`? -- so when other Presto developers search for `TaskSource`, they don't need to worry about "SerializedTaskSource":

<img width="666" alt="Screen Shot 2020-06-07 at 12 10 39 AM" src="https://user-images.githubusercontent.com/799346/83962583-59c42480-a853-11ea-9d7d-ef84f8528891.png">

wenleix commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -153,20 +152,10 @@ public PrestoSparkRddFactory(SplitManager splitManager, Metadata metadata, JsonC
-                    "Fragment with is not expected to have table scans. fragmentId: %s, fragment partitioning %s",
-                    fragment.getId(),
-                    fragment.getPartitioning());
+        // set's the number of output partitions
```
nit: do you mean `set the number of output partitions`?

wenleix commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
-            serializedTaskDescriptors.add(new SerializedPrestoSparkTaskDescriptor(jsonSerializedTaskDescriptor));
-            // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
-            assignedSplits.set(i, null);
+        IntStream partitions = expectedNumberOfPartitions
```
As you might guess, I will use more "plain" style `if`,`list`,`for-loop` for line 290-303 😃 . But it's personal taste.

wenleix replied:
Also, would it help understand code if bucketed table / unbucketed table split assignment is done in different code blocks? -- while they share the concept of "partitions", its semantic is different. For bucketed tables, you have to have the same number of partitions as with buckets, while for unbucketed tables, the number of partition is by configured or automatically tuned (e.g. based on table size).

arhimondr replied:
I actually wanted to use `IntStream` specifically to avoid materializing `IntStream.range(0, integer)`

wenleix replied:
> I actually wanted to use IntStream specifically to avoid materializing IntStream.range(0, integer)

 Why not just using for loop? ...

arhimondr replied:
You will have to loop either over the partitions from `taskSourcesMap` or over all partitions provided by `expectedNumberOfPartitions `. With simple `for` loop it seems like you would have to duplicate the loop body twice.

wenleix replied:
@arhimondr : Stream API often yields non-trivial performance overhead (see, for example https://github.com/prestodb/presto/pull/11374) . And I don't think this part of code is performance sensitive? (it's just doing plan once on Driver?)

I personally prefer not using Stream API unless it's pretty clear (e.g. the classic `.map().filter().XXX` pattern. This has also been discussed in https://github.com/prestodb/presto/pull/11374#issuecomment-418464932 . And I also have expressed similar ideas in my other code reviews. 

I understand a general policy discussion about Stream API should be done by TSC. 

In this specific case, I found it make hard for me to follow the core logic with Stream API 😕

wenleix replied:
@arhimondr : Java Stream API can also have non-trivial memory allocation overhead, see https://github.com/prestodb/presto/pull/13984

arhimondr replied:
Replaces with 2 branches and regular loops. Added comments.

wenleix commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
+
+        List<List<SerializedTaskSource>> partitionedTaskSources = new ArrayList<>();
+        partitions.forEach(partition -> {
+            List<TaskSource> taskSources = taskSourcesMap.removeAll(partition);
```
Add a comment explain why eagerly remove the task sources

arhimondr replied:
Added a comment

wenleix commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -330,49 +324,24 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
     }
 
-    private static List<List<ScheduledSplit>> assignSplitsToTasks(List<ScheduledSplit> splits, int numTasks)
+    private SetMultimap<Integer, ScheduledSplit> assignSplitsToTasks(Session session, PartitioningHandle partitioning, List<ScheduledSplit> splits)
```
curious: why using `SetMultimap` instead of `ListMultimap`?

arhimondr replied:
TaskSource takes `Set<ScheduledSplit> splits`

wenleix commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
+            List<ScheduledSplit> scheduledSplits = getSplits(session, tableScan);
+            shuffle(scheduledSplits);
+            SetMultimap<Integer, ScheduledSplit> assignedSplits = assignSplitsToTasks(session, partitioning, scheduledSplits);
+            asMap(assignedSplits).forEach((partitionId, splits) ->
```
nit: why not just using a `for-each` loop? ;)

arhimondr replied:
I like this style as you can give a key and a value a meaningful name

> (partitionId, splits) ->

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskProcessor.java:
```diff
@@ -0,0 +1,74 @@
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public class PrestoSparkTaskProcessor
```
So now the anonymous lambda created in legacy `TaskProcessors` now has dedicated named class, this is great. 

Curious why this class has to stay in `classloader_interface` ?

arhimondr replied:
It has to be serializable. If it is loaded not by the Spark class loader it cannot get serialized =\

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskProcessor.java:
```diff
@@ -0,0 +1,74 @@
+        this.broadcastInputs = unmodifiableMap(new HashMap<>(requireNonNull(broadcastInputs, "broadcastInputs is null")));
+    }
+
+    public Iterator<Tuple2<Integer, PrestoSparkRow>> apply(
```
maybe call it `process`? :)

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskProcessor.java:
```diff
@@ -0,0 +1,74 @@
+        int partitionId = TaskContext.get().partitionId();
+        int attemptNumber = TaskContext.get().attemptNumber();
+        Map<String, Iterator<Tuple2<Integer, PrestoSparkRow>>> inputsMap = new HashMap<>();
+        for (int i = 0; i < fragmentIds.size(); i++) {
```
should there be a check that `fragmentIds.size() == inputs.size()` ?

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskSourceRdd.java:
```diff
@@ -0,0 +1,86 @@
+import static java.util.stream.Collectors.toList;
+import static scala.collection.JavaConversions.asScalaBuffer;
+
+public class PrestoSparkTaskSourceRdd
```
For now Looks like it's a wrapper over `ParallelCollectionPartition` . I assume it will have non-trivial functionality when colocated join is supported :)

wenleix replied:
In fact I need to take a closer look into this. Looks like it's more than just a wrapper :) (especially this `getInputIterators` method)

arhimondr replied:
> For now Looks like it's a wrapper over ParallelCollectionPartition . I assume it will have non-trivial functionality when colocated join is supported :)

Yeah. It provides us with a flexibility to manually assign `TaskSource` to spark partitions.  `ParallelRdd` (the result of `parallelize`) just does round robin.

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskRdd.java:
```diff
@@ -0,0 +1,126 @@
+import static scala.collection.JavaConversions.asScalaIterator;
+import static scala.collection.JavaConversions.seqAsJavaList;
+
+public class PrestoSparkTaskRdd
```
Let's add some comment for this class. Feel free to adapt/modify the following:

```
/**
 * PrestoSparkTaskRdd represents execution of Presto stage, it contains:
 *  - a list of shuffleInputRdds, each of the corresponding to a child stage.
 *  - an optional taskSourceRdd, which represents *ALL* the table scan inputs in this stage.
 *
 *  Table scan presents when join bucketed table with unbucketed table, for example:
 *        Join
 *        /  \
 *    Scan  Remote
 *  In this case, bucket to Spark partition mapping has to be consistent with the Spark shuffle partition.
 *
 *  shuffleInputRdds can also be empty when the stage partitioning is SINGLE_DISTRIBUTION.
 *  TODO: Consider have dedicated RDD implementation for SINGLE_DISTRIBUTION.
 *
 *  The broadcast input is encapsulated in taskProcessor.
 */
```

Let me know if the TODO suggestion (TODO: Consider have dedicated RDD implementation for SINGLE_DISTRIBUTION) makes sense. Since I personally found it difficult to reason when both `shuffleInputRdds` and `taskSourceRdd` can be empty.

arhimondr replied:
> (TODO: Consider have dedicated RDD implementation for SINGLE_DISTRIBUTION) 

I'm not particularly convinced that a separate RDD specifically for single distribution makes a lot of sense, especially if sometimes single distributed fragment may still have inputs. It might make things even more confusing.

wenleix replied:
@arhimondr : Yeah, let's make how to handle `SINGLE_DISTRIBUTION` as a future work :)

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskRdd.java:
```diff
@@ -0,0 +1,126 @@
+        if (inputRdds.isEmpty()) {
+            return emptyList();
+        }
+        int startIndex = taskSourceRdd != null ? 1 : 0;
```
Here a implicit "sublist" is done and makes the code a bit difficult to follow (similar argument to `getTaskSourceIterator`). Would it make sense to do the `partitions.subList(1, ...)` in `compute`?

arhimondr replied:
I remember I was trying to do something like that at the beginning, but it was still messy + it was more code. I would prefer to keep it as is for now. Let me know what you think?

wenleix replied:
@arhimondr : I see. Maybe add a TODO about refactor and maybe I can look into that later? :)

arhimondr replied:
Let me try to store `taskSourceRdd ` as a last element. It might make things a little bit less confusing.

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskProcessor.java:
```diff
@@ -0,0 +1,74 @@
+
+    public Iterator<Tuple2<Integer, PrestoSparkRow>> apply(
+            Iterator<SerializedTaskSource> serializedTaskSources,
+            List<Iterator<Tuple2<Integer, PrestoSparkRow>>> inputs)
```
nit: `shuffleInputs`?

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskProcessor.java:
```diff
@@ -0,0 +1,74 @@
+    {
+        int partitionId = TaskContext.get().partitionId();
+        int attemptNumber = TaskContext.get().attemptNumber();
+        Map<String, Iterator<Tuple2<Integer, PrestoSparkRow>>> inputsMap = new HashMap<>();
```
nit: `shuffleInputsMap` or `fragmentIdToShuffleInput`?

wenleix replied:
So essentially, these two information are maintained in different classes:
- `fragmentIds` list is maintained in `PrestoSparkTaskProcessor`
- `shuffleInput` list is maintained in `PrestoSparkTaskRdd`.

And these two lists are "zipped" when `PrestoSparkTaskRdd#compute` calls `PrestoSparkTaskProcessor#apply`. 

I am wondering if this "zipping" operation can be done in `PrestoSparkTaskRdd#compute`? -- Otherwise, it's a bit uneasy to reason the ordering of `shuffleInput` in `PrestoSparkTaskRdd` has to be consistent with the ordering of `fragmentIds` in `PrestoSparkTaskProcessor`... 

When I am reading the code, I have to traverse the code in the following way:
-  Oh, how to guarantee the ordering in this two list are consistent?
- ok , `shuffleInput` is from `PrestoSparkTaskRdd`, and it's from `PrestoSparkRddFactory`
- And  the same `PrestoSparkRddFactory` provides the fragment id list when construct `PrestoSparkTaskProcessor`
- OK so these two lists are consistent...

<img width="140" alt="Screen Shot 2020-06-08 at 11 14 12 PM" src="https://user-images.githubusercontent.com/799346/84112540-ccf9a200-a9dd-11ea-842c-d29ea6f1ebea.png">

arhimondr replied:
> I am wondering if this "zipping" operation can be done in PrestoSparkTaskRdd#compute? 

Could you please elaborate a little bit more. I believe I'm missing something. The list of `shuffleInput ` is only available as a parameter of `process` in `PrestoSparkTaskProcessor`. What zipping you think can be done in `PrestoSparkTaskRdd`?

wenleix replied:
@arhimondr : 

> Could you please elaborate a little bit more. I believe I'm missing something. The list of shuffleInput is only available as a parameter of process in PrestoSparkTaskProcessor.

So essentially, you are zipping `fragmentIds` and `inputs` into `inputsMap`. And `fragmentIds` is a field in `PrestoSparkTaskProcessor`, while `inputs` is passed as a parameter into `PrestoSparkTaskProcessor#process`

arhimondr replied:
Yup

wenleix replied:
@arhimondr : 

> Yup

But it's not fixed. I mean, the "zipped inputs" are from two different source (one stay in `PrestoSparkTaskProcessor` as a field member, one provided through `PrestoSparkTaskProcessor#process`). And it's quite confusing. For example, can they both be provided from `PrestoSparkTaskProcessor#process`? 

I would leave this as a future discussion for now.

arhimondr replied:
Nevermind, I think I got your point now. Refactored. Good catch.

wenleix commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
-            assignedSplits.set(i, null);
+        IntStream partitions = expectedNumberOfPartitions
+                .map(integer -> IntStream.range(0, integer))
+                .orElseGet(() -> new ArrayList<>(taskSourcesMap.keySet()).stream().mapToInt(Integer::intValue));
```
Two questions: 

- Is it possible some partition is missing from the `taskSourcesMap`? (e.g. table contains empty bucket)
- Will it happen that `expectedNumberOfPartitions` is present but different from the number of partitions in `taskSourcesMap`?

arhimondr replied:
> Is it possible some partition is missing from the taskSourcesMap? (e.g. table contains empty bucket)

Yes, it is possible. `taskSourcesMap.removeAll(partition)` would return an empty list then. Even if the splits are empty, we still need to provide a partition, as the number of partitions should be the same for all the inputs that are zipped for consistency. 

> Will it happen that expectedNumberOfPartitions is present but different from the number of partitions in taskSourcesMap?

If it happens it means that the fragmented plan is wrong. The partitioning of the fragment should be consistent with partitioning of the TableScan.

wenleix commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
+        SerializedPrestoSparkTaskDescriptor serializedTaskDescriptor = new SerializedPrestoSparkTaskDescriptor(
+                taskDescriptorJsonCodec.toJsonBytes(taskDescriptor));
 
+        Optional<Integer> numberOfInputPartitions = Optional.empty();
```
Curious what does this `numberOfInputPartitions` mean? -- Looks like it's decided by the partition number of other RDDs?

arhimondr replied:
It represents number of partitions that are expected by the rdd based on it's inputs. The number of partitions should be equal for all input rdds. If there's no shuffle inputs we don't have to create empty partitions if the splits are missing for some buckets. Think of it as of an optimization for a query like

```
select col, count(*)
from table_bucketed_by_col
group by col
```

If some buckets are missing we are free to not to create tasks for buckets that are missing.

But for a query like 

```
select col, count(*)
from table_bucketed_by_col t1
RIGHT OUTER JOIN non_bucketed_table t2
WHERE t1.col = t2.col
```

even if some buckets are missing for t1 we still need to create a task for correctness

wenleix replied:
@arhimondr : I see. So essentially `numberOfInputPartitions` comes from the left-most join table right ?

i.e. considering 
```
            JOIN
          /   |   \
        A    B    C
```

Once `rddA` is created, the number of partitions in `rddA` will be populated to `B` and `C` for the number of partitions?

arhimondr replied:
> I see. So essentially numberOfInputPartitions comes from the left-most join table right ?

No. It should be determined by the partitions produced by shuffle.

```
           JOIN
          /   |   \
        A    B    C
```

If table A and C are bucketed, but table B is non bucketed - we will have to shuffle table B. Thus the number of partitions will be present, and will be set to the number of partitions produced by the shuffle over B.

If all tables A, B and C are bucketed, and if bucket 0 is missing for all of them - there's no need to schedule a task for that bucket. This optimization cannot be done if at least one table is non bucketed, as there's no way to know in advance if there will be any rows for that bucket produced by the shuffle.

wenleix replied:
@arhimondr : Maybe add some comments about the intention of`expectedNumberOfPartitions` (used to eliminate tasks when there are missing buckets), and its behavior when there are both bucketed table input and shuffle input, since it's not intuitive at first glance :)

wenleix commented on presto-spark-base/src/test/java/com/facebook/presto/spark/TestPrestoSparkQueryRunner.java:
```diff
@@ -55,20 +55,77 @@ public void testTableWrite()
+                "ON l.orderkey = o.orderkey " +
+                "WHERE l.orderkey % 223 = 42 AND l.linenumber = 4 and o.orderstatus = 'O'");
+
+        // only one table is bucketed
```
nit: maybe say "only probe side table is bucketed"

wenleix commented on presto-spark-base/src/test/java/com/facebook/presto/spark/TestPrestoSparkQueryRunner.java:
```diff
@@ -55,20 +55,77 @@ public void testTableWrite()
                 "ON l.orderkey = o.orderkey " +
                 "WHERE l.orderkey % 223 = 42 AND l.linenumber = 4 and o.orderstatus = 'O'");
+
+        // only one table is bucketed
```
ditto, "only build side table is bucketed"

arhimondr commented:
@wenleix 

> "Add ConnectorNodePartitioningProvider#getBucketCount". LGTM % one minor question. Also, does it make sense to add getBucketCount into ConnectorPartitioningHandle interface? (instead of putting in ConnectorNodePartitioningProvider) :) i.e.

Originally the Handles were designed as opaque marker interfaces. Over the time we've seen more and more utility methods added to handles that I personally not a very big fan of. I would prefer not to extend the Handle interfaces if there's no strong need for it. If you don't mind I would like to keep the existing approach.

arhimondr commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -230,90 +230,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
     {
         checkInputs(fragment.getRemoteSourceNodes(), rddInputs, broadcastInputs);
```
I don't feel particularly strong. I can do this rename in a separate PR if you wish, it's been like that before.

wenleix replied:
@arhimondr : Previously there is only one type of input (shuffle input). Now with two types of inputs, just saying `rddInputs` can be confusing when reading the code. 

Feel free to to it in a separate PR.

arhimondr commented on presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:
```diff
@@ -230,90 +230,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
+        }
+        else if (rddInputs.size() == 0) {
+            checkArgument(fragment.getPartitioning().equals(SINGLE_DISTRIBUTION), "SINGLE_DISTRIBUTION partitioning is expected: %s", fragment.getPartitioning());
+            taskSourceRdd = Optional.of(new PrestoSparkTaskSourceRdd(sparkContext.sc(), ImmutableList.of(ImmutableList.of())));
```
We need to somehow create exactly 1 partition, so exactly one task is scheduled. Thats why we need this single partition input RDD. I agree that this is a bit of a hack, and in theory it is possible to create a single "dummy" partition of the `PrestoSparkTaskRdd` itself. But that might be a little messy, so I'm kinda trying to chose a lesser evil.

wenleix replied:
@arhimondr : If the RDD becomes special enough in the case of `SINGLE_DISTRIBUTION` , a dedicated RDD might not be a bad idea, but you might have different opinions :) .

In any case, for now just explain in comment when the list is empty, it means exactly 1 partition (this is not intuitive to me), and perhaps a TODO about figure out less hacky approach? (We have 20% time assigned to Better Engineering).

Still , what about have a factory method such as `PrestoSparkTaskSourceRdd.createRddWithSingleDistribution` so you hide the detail that creating a `PrestoSparkTaskSourceRdd` with `ImmutableList.of(ImmutableList.of())` as task source?

arhimondr replied:
> @arhimondr : If the RDD becomes special enough in the case of SINGLE_DISTRIBUTION , a dedicated RDD might not be a bad idea, but you might have different opinions :) .

I would prefer to have a single way of creating a presto task. Otherwise we would have to maintain multiple ways. I think that's the main motivation to stick with  only `PrestoSparkTaskSourceRdd `, `PrestoSparkTaskRdd`.

> PrestoSparkTaskSourceRdd.createRddWithSingleDistribution

Then maybe `createRddWithSingleEmptyPartition`. But that might also be quite confusing. Let me leave a comment instead.

arhimondr commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskRdd.java:
```diff
@@ -0,0 +1,126 @@
+public class PrestoSparkTaskRdd
+        extends ZippedPartitionsBaseRDD<Tuple2<Integer, PrestoSparkRow>>
+{
+    private RDD<SerializedTaskSource> taskSourceRdd;
```
Good point

arhimondr commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskSourceRdd.java:
```diff
@@ -0,0 +1,86 @@
+public class PrestoSparkTaskSourceRdd
+        extends RDD<SerializedTaskSource>
+{
+    private List<List<SerializedTaskSource>> partitionedTaskSources;
```
Renames to `taskSourcesByPartitionId`, it feels like it fits better, as then in the `PrestoSparkTaskSourceRdd` we are creating partitions and iterating over partitions.

Added a comment as well.

arhimondr commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskRdd.java:
```diff
@@ -0,0 +1,126 @@
+    private Iterator<SerializedTaskSource> getTaskSourceIterator(List<Partition> partitions, TaskContext context)
+    {
+        if (taskSourceRdd != null) {
+            return asJavaIterator(taskSourceRdd.iterator(partitions.get(0), context));
```
Yeah, generally this class is a little messy, as it has to multiplex both, inputs with splits with inputs with rows. I was trying to meddle it around, but it seems that it is getting even worse.

wenleix replied:
@arhimondr : What about the second option (let `compute` to decide whether to pass in `Optional.empty()` or `partitions.get(0)`

arhimondr replied:
Actually let me change this code to store source partitions as a last element. It might be easier to read it then.

wenleix commented:
Some high level thoughts:

- Would be great to have @viczhang861  or @rschlussel  to also take a look at the new split assignment part (basically `PrestoSparkRddFactory#createTaskSourcesRdd` and utility methods such as `PrestoSparkRddFactory#assignSplitsToTasks`, `PrestoSparkRddFactory#assignSourceDistributionSplits`, `PrestoSparkRddFactory#assignPartitionedSplits`.

- In my experience there is a gap between `TestAbstractDistributedQueries` and running production queries,  due to the following reasons:
      1. `TestAbstractDistributedQueries` doesn't contain INSERT (Read-only queries)
      2. `TestAbstractDistributedQueries` doesn't test bucket table.

    That's why `HiveIntegrationSmokeTest` contains a lot of queries not covered by `TestAbstractDistributedQueries` (corner cases for bucket join/aggregation/union, empty bucket, etc, despite it's originally designed to be a smoke test). As a future work we might want to migrate some of such tests into `TestPrestoSparkQueryRunner`

- We should do some production tests to make sure it works at scale after so many changes (especially being able to directly using a fat executor).

arhimondr commented:
> We should do some production tests to make sure it works at scale after so many changes

I was running production test recently. It seemed to be working fine.

wenleix commented on presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskRdd.java:
```diff
@@ -0,0 +1,120 @@
+ * <p>
+ * In this case, bucket to Spark partition mapping has to be consistent with the Spark shuffle partition.
+ * <p>
+ * shuffleInputRdds can also be empty when the stage partitioning is SINGLE_DISTRIBUTION.
```
Maybe emphasize both `taskSourceRdd` and `shuffleInputRdds` will be empty when the stage partitioning is `SINGLE_DISTRIBUTION` ?

arhimondr replied:
Changed the comment to 

```
 * When the stage partitioning is SINGLE_DISTRIBUTION and the shuffleInputRdds is empty,
 * the taskSourceRdd is expected to be present and contain exactly one empty partition.
```

Ravion commented:
Hi,

Is there any documents for configuring and running Presto on Spark? Thanks

Regards,
Ravi

On Mon, Jun 15, 2020, 12:37 PM Andrii Rosa <notifications@github.com> wrote:

> *@arhimondr* commented on this pull request.
> ------------------------------
>
> In
> presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java
> <https://github.com/prestodb/presto/pull/14559#discussion_r440304076>:
>
> >          }
>
> -        List<List<ScheduledSplit>> assignedSplits = assignSplitsToTasks(splits, numTasks);
> -
> -        // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
> -        splits = null;
> -
> -        ImmutableList.Builder<SerializedPrestoSparkTaskDescriptor> serializedTaskDescriptors = ImmutableList.builder();
> -        for (int i = 0; i < assignedSplits.size(); i++) {
> -            List<ScheduledSplit> splitBatch = assignedSplits.get(i);
> -            PrestoSparkTaskDescriptor taskDescriptor = createSourceTaskDescriptor(session, tableWriteInfo, fragment, splitBatch);
> -            // TODO: consider more efficient serialization or apply compression to save precious memory on the Driver
> -            byte[] jsonSerializedTaskDescriptor = taskDescriptorJsonCodec.toJsonBytes(taskDescriptor);
> -            serializedTaskDescriptors.add(new SerializedPrestoSparkTaskDescriptor(jsonSerializedTaskDescriptor));
> -            // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
> -            assignedSplits.set(i, null);
> +        IntStream partitions = expectedNumberOfPartitions
>
> Replaces with 2 branches and regular loops. Added comments.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/14559#discussion_r440304076>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3TLDDARZQUXVP7GBFFBA3RWZE3FANCNFSM4NFLXYZA>
> .
>

arhimondr commented:
@wenleix 

> A specific topic is about Java Stream API: #14559 (comment)

Changed to a simple loop, added more comments

> PrestoSparkTaskProcessor#process has input parameters from very different source

Fixed

> Avoid creating task in the presence of missing bucket:

Added a comment

arhimondr commented:
Hi @Ravion!

Unfortunately there's no documentation yet. 

But running Presto on Spark is simple. Basically you build the project with

`./mvnw clean install -DskipTests -TC1`

And then you run spark-submit providing the artifacts built

e.g.: 
```
/spark/bin/spark-submit \
    --master spark://spark-master:7077 \
    --executor-cores 4 \
    --conf spark.task.cpus=4 \
    --class com.facebook.presto.spark.launcher.PrestoSparkLauncher \
    /presto/presto-spark-launcher/target/presto-spark-launcher-*-SNAPSHOT.jar \
    --package /presto/presto-spark-package/target/presto-spark-package-*-SNAPSHOT.tar.gz \
    --config /presto/etc/config.properties \
    --catalogs /presto/etc/catalogs \
    --catalog hive \
    --schema default \
    --file /tmp/query.sql
```

Example config.properties

```
query.hash-partition-count=10
redistribute-writes=false
#join-distribution-type=BROADCAST
```

Example `catalogs/hive.properties`

```
connector.name=hive-hadoop2
hive.dfs.require-hadoop-native=false
hive.time-zone=America/Bahia_Banderas
hive.metastore.uri=thrift\://hadoop-master\:9083
```

You can try to run it and let us know how it goes. Happy to answer any follow up questions.

Ravion commented:
Thanks..Much appreciate the quick response.

Best,
Ravi

On Mon, Jun 15, 2020 at 1:21 PM Andrii Rosa <notifications@github.com>
wrote:

> Hi @Ravion <https://github.com/Ravion>!
>
> Unfortunately there's no documentation yet.
>
> But running Presto on Spark is simple. Basically you build the project with
>
> ./mvnw clean install -DskipTests -TC1
>
> And then you run spark-submit providing the artifacts built
>
> e.g.:
>
> /spark/bin/spark-submit \
>     --master spark://spark-master:7077 \
>     --executor-cores 4 \
>     --conf spark.task.cpus=4 \
>     --class com.facebook.presto.spark.launcher.PrestoSparkLauncher \
>     /presto/presto-spark-launcher/target/presto-spark-launcher-*-SNAPSHOT.jar \
>     --package /presto/presto-spark-package/target/presto-spark-package-*-SNAPSHOT.tar.gz \
>     --config /presto/etc/config.properties \
>     --catalogs /presto/etc/catalogs \
>     --catalog hive \
>     --schema default \
>     --file /tmp/query.sql
>
> Example config.properties
>
> query.hash-partition-count=10
> redistribute-writes=false
> #join-distribution-type=BROADCAST
>
> Example catalogs/hive.properties
>
> connector.name=hive-hadoop2
> hive.dfs.require-hadoop-native=false
> hive.time-zone=America/Bahia_Banderas
> hive.metastore.uri=thrift\://hadoop-master\:9083
>
> You can try to run it and let us know how it goes. Happy to answer any
> follow up questions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/14559#issuecomment-644264393>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3TLDAYX7SBO74PDI3D7ADRWZKB5ANCNFSM4NFLXYZA>
> .
>

Ravion commented:
Hi @Andril Rosa,

Thanks for the inputs. Perfectly works. I will write an article and will
send the link. Is it possible to use the approach for interactive querying,
or is it that we need to give SQL in a file always ?

Screenshot is attached.

Best,
Ravi

On Mon, Jun 15, 2020 at 1:21 PM Andrii Rosa <notifications@github.com>
wrote:

> Hi @Ravion <https://github.com/Ravion>!
>
> Unfortunately there's no documentation yet.
>
> But running Presto on Spark is simple. Basically you build the project with
>
> ./mvnw clean install -DskipTests -TC1
>
> And then you run spark-submit providing the artifacts built
>
> e.g.:
>
> /spark/bin/spark-submit \
>     --master spark://spark-master:7077 \
>     --executor-cores 4 \
>     --conf spark.task.cpus=4 \
>     --class com.facebook.presto.spark.launcher.PrestoSparkLauncher \
>     /presto/presto-spark-launcher/target/presto-spark-launcher-*-SNAPSHOT.jar \
>     --package /presto/presto-spark-package/target/presto-spark-package-*-SNAPSHOT.tar.gz \
>     --config /presto/etc/config.properties \
>     --catalogs /presto/etc/catalogs \
>     --catalog hive \
>     --schema default \
>     --file /tmp/query.sql
>
> Example config.properties
>
> query.hash-partition-count=10
> redistribute-writes=false
> #join-distribution-type=BROADCAST
>
> Example catalogs/hive.properties
>
> connector.name=hive-hadoop2
> hive.dfs.require-hadoop-native=false
> hive.time-zone=America/Bahia_Banderas
> hive.metastore.uri=thrift\://hadoop-master\:9083
>
> You can try to run it and let us know how it goes. Happy to answer any
> follow up questions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/14559#issuecomment-644264393>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3TLDAYX7SBO74PDI3D7ADRWZKB5ANCNFSM4NFLXYZA>
> .
>

arhimondr commented:
Hi @Ravion!

> Is it possible to use the approach for interactive querying,
or is it that we need to give SQL in a file always ?

Currently that's the only supported interface. We recommend using classic Presto for interactive querying. Presto on Spark is designed to run long running batch (ETL) workload.

Ravion commented:
Got it, makes sense...Good job on this.

Best, Ravi

On Tue, Jun 16, 2020, 1:24 PM Andrii Rosa <notifications@github.com> wrote:

> Hi @Ravion <https://github.com/Ravion>!
>
> Is it possible to use the approach for interactive querying,
> or is it that we need to give SQL in a file always ?
>
> Currently that's the only supported interface. We recommend using classic
> Presto for interactive querying. Presto on Spark is designed to run long
> running batch (ETL) workload.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/14559#issuecomment-644902270>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3TLDHXOOC4JOBAOIQCTXTRW6TDHANCNFSM4NFLXYZA>
> .
>
